### PR TITLE
[RISCV] fix SP recovery in a function epilogue

### PR DESCRIPTION
**Pull request by daniilavdeev**
Currently, in the cases when fp register is presented and sp register is adjusted at the second time, sp recovery in a function epilogue isn't performed in the best way, for example:
```
lui a0, 2
sub sp, s0, a0
addi a0, a0, -2044
add sp, sp, a0
```

This patch improves sp register recovery in such cases and the code snippet above becomes:
```
addi sp, s0, -2044
```

### CODE REVIEW

**Comment by dlav-sc** [comment]
@topperc @kito-cheng FYI

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 1]
```diff
@@ -786,74 +799,90 @@ void RISCVFrameLowering::emitEpilogue(MachineFunction &MF,
...
-  //
-  // In order to make sure the stack point is right through the EH region,
-  // we also need to restore stack pointer from the frame pointer if we
-  // don't preserve stack space within prologue/epilogue for outgoing variables,
-  // normally it's just checking the variable sized object is present or not
-  // is enough, but we also don't preserve that at prologue/epilogue when
-  // have vector objects in stack.
-  if (RI->hasStackRealignment(MF) || MFI.hasVarSizedObjects() ||
-      !hasReservedCallFrame(MF)) {
-    assert(hasFP(MF) && "frame pointer should not have been eliminated");
```
Comment (kito-cheng) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:815]: Could you try to keep this assertion in some where if possible?

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 2]
```diff
@@ -1566,6 +1595,7 @@ void RISCVFrameLowering::emitCalleeSavedRVVPrologCFI(
     int FI = CS.getFrameIdx();
     if (FI >= 0 && MFI.getStackID(FI) == TargetStackID::ScalableVector) {
       MCRegister BaseReg = TRI.getSubReg(CS.getReg(), RISCV::sub_vrm1_0);
+
```
[the remark this answers is not in the document]
Comment (dlav-sc) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp]: addressed

File: llvm/lib/Target/RISCV/RISCVFrameLowering.h [hunk 3]
```diff
@@ -88,9 +88,15 @@ class RISCVFrameLowering : public TargetFrameLowering {
   void adjustStackForRVV(MachineFunction &MF, MachineBasicBlock &MBB,
                          MachineBasicBlock::iterator MBBI, const DebugLoc &DL,
                          int64_t Amount, MachineInstr::MIFlag Flag) const;
+
```
[the remark this answers is not in the document]
Comment (dlav-sc) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.h]: addressed

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp (hunk 1 above)
Comment (dlav-sc) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:815]: addressed

**Review by kito-cheng — APPROVED** [review]
LGTM :)

**Comment by topperc** [comment]
This patch needs a better description. Is it fixing a correctness issue or just making code more optimal?

**Comment by dlav-sc** [comment]
> This patch needs a better description. Is it fixing a correctness issue or just making code more optimal?

I've updated the description

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 4]
```diff
@@ -786,20 +799,51 @@ void RISCVFrameLowering::emitEpilogue(MachineFunction &MF,
...
+  uint64_t FPOffset = FirstSPAdjustAmount ? FirstSPAdjustAmount
+                                          : getStackSizeWithRVVPadding(MF) -
+                                                RVFI->getVarArgsSaveSize();
   uint64_t RVVStackSize = RVFI->getRVVStackSize();
 
+  bool RestoreFP = RI->hasStackRealignment(MF) || MFI.hasVarSizedObjects() ||
+                   !hasReservedCallFrame(MF);
+
+  if (RVVStackSize) {
+    // If restoreFP the stack pointer will be restored using the frame pointer
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:824]: restoreFP -> RestoreFP

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 5]
```diff
@@ -786,20 +799,51 @@ void RISCVFrameLowering::emitEpilogue(MachineFunction &MF,
...
+    }
+  }
+
+  if (FirstSPAdjustAmount) {
+    uint64_t SecondSPAdjustAmount =
+        getStackSizeWithRVVPadding(MF) - FirstSPAdjustAmount;
+    assert(SecondSPAdjustAmount > 0 &&
+           "SecondSPAdjustAmount should be greater than zero");
+
+    // If restoreFP the stack pointer will be restored using the frame pointer
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:838]: restoreFP -> RestoreFP

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 6]
```diff
@@ -786,20 +799,51 @@ void RISCVFrameLowering::emitEpilogue(MachineFunction &MF,
...
+                                                RVFI->getVarArgsSaveSize();
   uint64_t RVVStackSize = RVFI->getRVVStackSize();
 
+  bool RestoreFP = RI->hasStackRealignment(MF) || MFI.hasVarSizedObjects() ||
+                   !hasReservedCallFrame(MF);
+
+  if (RVVStackSize) {
+    // If restoreFP the stack pointer will be restored using the frame pointer
+    // value.
+    if (!RestoreFP) {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:826]: Drop curly braces

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 7]
```diff
@@ -786,20 +799,51 @@ void RISCVFrameLowering::emitEpilogue(MachineFunction &MF,
...
+
+  if (FirstSPAdjustAmount) {
+    uint64_t SecondSPAdjustAmount =
+        getStackSizeWithRVVPadding(MF) - FirstSPAdjustAmount;
+    assert(SecondSPAdjustAmount > 0 &&
+           "SecondSPAdjustAmount should be greater than zero");
+
+    // If restoreFP the stack pointer will be restored using the frame pointer
+    // value.
+    if (!RestoreFP) {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:840]: Drop curly braces

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp (hunk 4 above)
Comment (dlav-sc) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp]: addressed

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp (hunk 5 above)
Comment (dlav-sc) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp]: addressed

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp (hunk 7 above)
Comment (dlav-sc) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp]: addressed

**Review by topperc — APPROVED** [review]
LGTM

**Comment by dlav-sc** [comment]
Thanks for taking a look at the patch

**Comment by bscarlet** [comment]
I believe this change is the cause of a miscompilation I'm seeing (automatically linked above). Please take a look, and kindly revert the change until that issue is resolved.

**Comment by joanahalili** [comment]
Hello, 
This commit is causing a large number of targets failing on our end, for which there is also a reproducer posted on the issue above https://github.com/llvm/llvm-project/issues/113488.  Could you please have a look and possibly revert?
If there is no activity on the commit we will have to revert ourselves to unblock us.

**Comment by dlav-sc** [comment]
@bscarlet, @joanahalili 
Hi, I couldn't reproduce the issue.

Here is what I've tried to do:

1. clone https://github.com/llvm/llvm-project.git
2. git checkout 7be2ce7312191894c2458ecc08de7683fd2314fb
3. compile your program, save resulting assembly
4. git revert 7be2ce7312191894c2458ecc08de7683fd2314fb
4. compile your program, save resulting assembly
5. in both instances the code is the same: [assembly.txt](https://github.com/user-attachments/files/17542657/assembly.txt)

It does not look like my patch is the reason for the failure. Besides,
my patch should not even touch function prologue.

Could you please elaborate how can I reproduce the issue using existing codebase?
Or is my attempt to reproduce incorrect?

**Comment by bscarlet** [comment]
Could you please include the complete flags you use to build LLVM so that I can reproduce your entire procedure.

Did you use the flags I included in my reproduction case in your steps 3 & 5?

**Comment by dlav-sc** [comment]
Oh, I'm sorry, I've compiled your example using the wrong clang, my bad.

Anyway, I've received almost identical snippets:  [with_patch.txt](https://github.com/user-attachments/files/17545048/with_patch.txt), [without_patch.txt](https://github.com/user-attachments/files/17545073/without_patch.txt). In fact, they only differ in the clang commit hash, so you can ensure that the first one was compiled with my patch and the second one without it.

> Could you please include the complete flags you use to build LLVM so that I can reproduce your entire procedure.

```
$ cmake -S llvm -B build -GNinja -DLLVM_ENABLE_PROJECTS='clang' -DCMAKE_BUILD_TYPE=Release
$ cd build
$ cmake --build ./ -t clang
```

> Did you use the flags I included in my reproduction case in your steps 3 & 5?

Yep
```
$ bin/clang -S -O1 -mrvv-vector-bits=512 -march=rv64gcv1p0 --sysroot=<path>/<to>/riscv-gcc/sysroot --gcc-toolchain=<path>/<to>/riscv-gcc --target=riscv64-unknown-linux riscv-varargs-crash.cc -o riscv-varargs-crash.s
```

**Comment by bscarlet** [comment]
We can now confirm that despite initial appearances, this patch does not seem to be the culprit. I've closed the corresponding issue. Apologies for the false alarm, and thank you for helping investigate.

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 8]
```diff
@@ -786,20 +799,49 @@ void RISCVFrameLowering::emitEpilogue(MachineFunction &MF,
...
-  uint64_t RealStackSize = getStackSizeWithRVVPadding(MF);
-  uint64_t StackSize = RealStackSize - RVFI->getReservedSpillsSize();
-  uint64_t FPOffset = RealStackSize - RVFI->getVarArgsSaveSize();
+  uint64_t FirstSPAdjustAmount = getFirstSPAdjustAmount(MF);
+  uint64_t RealStackSize = FirstSPAdjustAmount ? FirstSPAdjustAmount
+                                               : getStackSizeWithRVVPadding(MF);
+  uint64_t StackSize = FirstSPAdjustAmount ? FirstSPAdjustAmount
+                                           : getStackSizeWithRVVPadding(MF) -
+                                                 RVFI->getReservedSpillsSize();
+  uint64_t FPOffset = FirstSPAdjustAmount ? FirstSPAdjustAmount
```
Comment (eaeltsin) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:815]: Can you please double-check that this line is correct?

When I change it to
```
uint64_t FPOffset = RealStackSize - RVFI->getVarArgsSaveSize();
```
our tests pass.

Comment (dlav-sc) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:815]: Could you provide a reproduction, please?

Could you follow the steps from my comment https://github.com/llvm/llvm-project/pull/110809#issuecomment-2441512277 and ensure that the first executable crashes and the second one doesn't.

**Comment by eaeltsin** [comment]
I'm not sure about crashing, but looking at disassembler at https://godbolt.org/z/z8Kehr5ob:

The new code for function foo has prologue
```
        addi    sp, sp, -496
        sd      ra, 424(sp)
        sd      s0, 416(sp)
        addi    s0, sp, 432
```
and epilogue
```
        addi    sp, s0, -496  <---- should be 432
        ld      ra, 424(sp)
        ld      s0, 416(sp)
        addi    sp, sp, 496
        ret
```

**Comment by dlav-sc** [comment]
> @dlav-sc is the above sufficient?

Yes, I've reproduced your problem and preparing a fix now, I hope I could open PR in an hour.

Thank you for your example, I'll make a test from it.

**Pull request merged**